### ISSUE TITLE
Add a PEP578 audit hook for Asyncio loop stalls

### ISSUE THREAD

**Issue by 92a9d474-8ead-41fd-ae55-aaa57da0cedd** [python/cpython#88241]
BPO | [44075](https://bugs.python.org/issue44075)
--- | :---
Nosy | @tiran, @asvetlov, @1st1, @zooba, @orf
PRs | <li>python/cpython#25990</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-05-08.12:19:46.330>
labels = ['3.11', 'type-feature', 'expert-asyncio']
title = 'Add a PEP578 audit hook for Asyncio loop stalls'
updated_at = <Date 2022-03-19.12:00:57.728>
user = 'https://github.com/orf'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-19.12:00:57.728>
actor = 'asvetlov'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2021-05-08.12:19:46.330>
creator = 'orf'
dependencies = []
files = []
hgrepos = []
issue_num = 44075
keywords = ['patch']
message_count = 6.0
messages = ['393251', '393253', '393255', '393263', '393267', '415542']
nosy_count = 5.0
nosy_names = ['christian.heimes', 'asvetlov', 'yselivanov', 'steve.dower', 'orf']
pr_nums = ['25990']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue44075'
versions = ['Python 3.11']
```

</p></details>

**Comment by kumaraditya303** [comment]
I agree with Andrew that audit events should not be used here. Also the change will slow down the eventloop as it need to constantly monitor the time. 

@gvanrossum What do you think?

**Comment by gvanrossum** [comment]
I presume this is about [this code](https://github.com/python/cpython/blob/9a11ed8e50492d327e4de0a8f3a473e788b14a6f/Lib/asyncio/base_events.py#L1911-L1923). (OP @orf, please confirm if you're still listening.)

I'm not sure I buy the reason why audit events are more desirable than using the logger -- the latter seems intended for these kinds of things, and it's logging at a pretty high level (warning, just short of error). But maybe we can make it so that you don't have to restart in debug mode, which I agree is annoying.

The key question is, can we afford to call `self.time()` twice for each `handle._run()` call? That seems excessive outside debug mode. Maybe we can time the [entire for-loop](https://github.com/python/cpython/blob/9a11ed8e50492d327e4de0a8f3a473e788b14a6f/Lib/asyncio/base_events.py#L1907) even outside debug mode? It might require tweaking the `slow_callback_duration` variable (currently set to 0.1 second), or having a separate variable?

**Comment by kumaraditya303** [comment]
@orf Are you still interested in this? If not then we can close this issue.

**Comment by orf** [comment]
Hey! I’m definitely still interested in the general idea of being able to accurately track loop stalls in production - to me this is absolutely critical.

We just use a variant of aiodebug, [which works by patching `Handle._run`](https://gitlab.com/quantlane/libs/aiodebug/-/blob/main/aiodebug/log_slow_callbacks.py#L25).

I do appreciate the discussion around this, but realistically I think the overhead of fetching the time or triggering an audit event is perfectly acceptable for real world applications that want to benefit from asyncio.

By not allowing this, or by locking it behind a “debug” flag with other overheads, we seem to be either targeting microbenchmarks or the elusive segment of production apps that are perfectly coded and so don’t require any kind of monitoring, but also require maximum throughput.

All this is to say, I don’t see how we *can’t* afford to pay this overhead in some way.

**Comment by orf** [comment]
Using the existing logger is one approach, but hooking it into something like datadog/stated is more effort than it should be, with a higher overhead.

I don’t know the right approach here to be honest. All I can phrase it as is “detecting and correlating loop stalls + performance issues in an automated way is more difficult than it could be”.

**Comment by gvanrossum** [comment]
What if we added a new API that allowed you to set a callback when a loop stall is detected? By default it would do nothing and skip the two time() calls (on my 3 year old Mac one monotonic() call takes around 100 nsec).

**Comment by orf** [comment]
I think a new API ion the loop for this would be perfect, and it’s a lot more explicit than a logger + an env var.

**Comment by gvanrossum** [comment]
Cool. Do you want to submit a PR? If you want to you we can first bikeshed the design a bit here (mostly the naming of the new method, and the signature of the callback, and whether there needs to be an API to get the current callback, etc. -- possibly following asyncio traditions).

**Comment by orf** [comment]
Sure! I’m going on holiday next week, but I’ll be able to work on a MR after.

One thing I wanted to consider was multiple callbacks. Should be constrain ourselves to a single callback? An integration library from datadog or Sentry would likely want to register their own callbacks here, which could cause conflicts.

**Comment by gvanrossum** [comment]
Yeah, that’s an eternal API design dilemma, it seems. We have this problem for signals. So let’s be robust and over-engineer it this time. :-)
